### PR TITLE
[DROOLS-1379] Add tests for OOPath in queries

### PR DESCRIPTION
**Pull request by jiripetrlik**
New tests for OOPath [DROOLS-1379]. These tests are focused on OOPath usage in queries. It also tests reactive and non-reactive behaviour of OOPath in queries.

### CODE REVIEW

File: drools-compiler/src/test/java/org/drools/compiler/oopath/OOPathQueriesTest.java [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    public void testQueryFromCode() {
+        final String drl =
+                "import org.drools.compiler.oopath.model.Thing;\n" +
+                        "query isContainedIn( Thing $x, Thing $y )\n" +
+                        "    $y := /$x/children\n" +
+                        "or\n" +
+                        "    ( $z := /$x/children and isContainedIn( $z, $y; ) )\n" +
+                        "end\n";
+
+        final List<String> itemList = Arrays.asList(new String[] { "display", "keyboard", "processor" });
```
Comment (triceo) [inline drools-compiler/src/test/java/org/drools/compiler/oopath/OOPathQueriesTest.java]: I suggest you pay more attention to variable locality. (Even if not literal.) `itemList` is first used on line 58, why is it declared on line 51? `ksession` is first used on line 63, why is it declared on line 53?

Don't force people to keep in their heads more than they need to, for longer than they need to.

File: drools-compiler/src/test/java/org/drools/compiler/oopath/OOPathQueriesTest.java [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        }
+
+        ksession.insert(smartphone);
+
+        QueryResults queryResults = ksession.getQueryResults("isContainedIn", new Object[] { smartphone, Variable.v });
+        List<String> resultList = new ArrayList<>();
+        for(QueryResultsRow row : queryResults) {
+            Thing component = (Thing) row.get("$y");
+            resultList.add(component.getName());
+        }
```
Comment (triceo) [inline drools-compiler/src/test/java/org/drools/compiler/oopath/OOPathQueriesTest.java]: This loop incl. the list initialization would look sooo nice as a `Stream`. ;-)

**Review by triceo — COMMENTED** [review]
Few stylistic comments, nothing serious.

File: drools-compiler/src/test/java/org/drools/compiler/oopath/OOPathQueriesTest.java [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+                        "    ( $z := /$x/children and isContainedIn( $z, $y; ) )\n" +
+                        "end\n";
+
+        final List<String> itemList = Arrays.asList(new String[] { "display", "keyboard", "processor" });
+
+        final KieSession ksession = new KieHelper().addContent( drl, ResourceType.DRL )
+                .build()
+                .newKieSession();
+
+        Thing smartphone = new Thing("smartphone");
```
Comment (baldimir) [inline drools-compiler/src/test/java/org/drools/compiler/oopath/OOPathQueriesTest.java]: Please make local variables final. Currently some of them are final and some not (I mean some of them that could be marked final are not). It's a bit confusing if it is mixed up like this.

File: drools-compiler/src/test/java/org/drools/compiler/oopath/OOPathQueriesTest.java [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+                    "end\n" +
+                    "rule \"Turn heating off\" when\n" +
+                    "    $r : Room()\n" +
+                    "    temperature( $r, $t; )\n" +
+                    "    eval( $t > 20 )" +
+                    "then\n" +
+                    "    $r.getHeating().setOn(false);\n" +
+                    "end\n";
+
+        Room room = new Room("Room");
```
Comment (baldimir) [inline drools-compiler/src/test/java/org/drools/compiler/oopath/OOPathQueriesTest.java]: This test method body looks the same in each test. Probably could be extracted to a method and reused.

**Review by baldimir — CHANGES_REQUESTED** [review]
Review done. Please see my and Lukas's comments. Only minor things.

**Comment by jiripetrlik** [comment]
I've improved PR according to suggestions.

**Review by baldimir — APPROVED** [review]
Looks good, +1.

**Review by triceo — APPROVED** [review]

File: drools-compiler/src/test/java/org/drools/compiler/oopath/OOPathQueriesTest.java [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+        final KieSession ksession = new KieHelper().addContent( drl, ResourceType.DRL )
+                .build()
+                .newKieSession();
+        ksession.insert(smartphone);
+
+        final QueryResults queryResults = ksession.getQueryResults("isContainedIn", new Object[] { smartphone, Variable.v });
+        final List<String> resultList = StreamSupport.stream(queryResults.spliterator(), false)
+                .map(row -> (Thing) row.get("$y")).map(thing -> thing.getName()).collect(Collectors.toList());
+        assertThat(resultList).as("Querry does not contain all items").containsAll(itemList);
```
Comment (yurloc) [inline drools-compiler/src/test/java/org/drools/compiler/oopath/OOPathQueriesTest.java]: Typo querry -> query.

File: drools-compiler/src/test/java/org/drools/compiler/oopath/OOPathQueriesTest.java [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        itemList.stream().map(item -> new Thing(item)).forEach((thing) -> smartphone.addChild(thing));
+
+        final KieSession ksession = new KieHelper().addContent( drl, ResourceType.DRL )
+                .build()
+                .newKieSession();
+        ksession.insert(smartphone);
+
+        final QueryResults queryResults = ksession.getQueryResults("isContainedIn", new Object[] { smartphone, Variable.v });
+        final List<String> resultList = StreamSupport.stream(queryResults.spliterator(), false)
+                .map(row -> (Thing) row.get("$y")).map(thing -> thing.getName()).collect(Collectors.toList());
```
Comment (yurloc) [inline drools-compiler/src/test/java/org/drools/compiler/oopath/OOPathQueriesTest.java]: Can be simplified to a single map call `.map(row -> ((Thing) row.get("$y")).getName())`.

**Comment by jiripetrlik** [comment]
Thank you @yurloc. Typo was fixed and stream expression was improved according to your suggestion.

**Comment by baldimir** [comment]
jenkins retest this

**Comment by mariofusco** [comment]
ok to test

**Comment by baldimir** [comment]
jenkins retest this

**Comment by mariofusco** [comment]
Replaced by https://github.com/kiegroup/drools/pull/1292 with 2 tests removed as explained in https://issues.jboss.org/browse/RHBRMS-2761

**Pull request closed without merging**